docs/cryptodoc/src/05_08_ml_dsa.rst

```diff
@@ -0,0 +1,443 @@
+      9. *Retry* iff :math:`\|z\|_{\infty} \geq \gamma_1 - \beta` (see ``infinity_norm_within_bound``)
+      10. :math:`cs_2 = NTT^{-1}(\hat{c} \cdot sk.\hat{s_2})`
+      11. :math:`r_0 = w_0 - cs_2`
+      12. *Retry* iff :math:`\|r_0\|_{\infty} \geq \gamma_2 - \beta` (see ``infinity_norm_within_bound``))
+      13. :math:`ct_0 = NTT^{-1}(\hat{c} \cdot sk.\hat{t_0})`
+      14. *Retry* iff :math:`\|c\|_{\infty} \geq \gamma_2` (see ``infinity_norm_within_bound``)
+      15. :math:`h = make\_hint(r_0 + ct_0, w_1)`
+      16. *Retry* if the Hamming weight of :math:`h > \omega`
+
+   4. :math:`\sigma = (\tilde{c}, z, h)` encoded using ``encode_signature``
+
+   **Notes:**
+
+   - This algorithm description assumes that the private signing key is expanded
+     into the internal representation already. Additionally, the expansion of
+     :math:`\hat{A}`, as well as the NTT for :math:`s_1`, :math:`s_2`, and
+     :math:`t_0` are done :srcref:`prior to the actual signing operation
+     <src/lib/pubkey/dilithium/dilithium_common/dilithium.cpp:131|Dilithium_Signature_Operation>`
+     to armortize the complexity of these operations across multiple consecutive
+     signature generations.
+   - Step 1: Botan 3.6.0 does not yet support the application-defined context
+     string as specified in [FIPS-204]_ Algorithm 2. See `GitHub #4376
+     <https://github.com/randombit/botan/issues/4376>`_.
+   - Step 3.12: We calculate :math:`r_0 = w_0 - cs_2`, this is equivalent to
+     :math:`r_0 = LowBits(w - cs_2)` as specified in [FIPS-204]_ Algorithm 7.
+   - Step 3.15: We generate the hint from :math:`w_0 - cs_2 + ct_0` this is equivalent
+     to the specification's :math:`w - cs_2 + ct_0`.
+
+.. _pubkey/ml_dsa/verification:
+
+Signature Verification
+----------------------
+
+Signature verification as specified in [FIPS-204]_ Algorithms 3 and 8 are
+implemented in :srcref:`Dilithium_Verification_Operation::is_valid_signature
+<src/lib/pubkey/dilithium/dilithium_common/dilithium.cpp:269|is_valid_signature>`
+with the preparation of the message representative :math:`\mu` being done in
+:srcref:`DilithiumMessageHash
+<src/lib/pubkey/dilithium/dilithium_common/dilithium_symmetric_primitives.h:31|DilithiumMessageHash>`.
+
+.. admonition:: Dilithium_Verification_Operation::is_valid_signature
+
+   **Input:**
+
+   - ``pk``: public verification key, with :math:`\hat{A}` and :math:`\hat{t_1'} = NTT(t_1 \cdot 2^{d})` in NTT domain
+   - ``M``: message to be verified
+   - ``signature``: the signature to be verified
+
+   **Output:**
+
+   - ``ok``: boolean value whether or not the signature is valid
+
+   **Steps:**
+
+   1. Calculate the message representative :math:`\mu = H(H(pk) \| 0x00 \| 0x00 \| M)` (see :srcref:`here <src/lib/pubkey/dilithium/ml_dsa/ml_dsa_impl.h:36|start>`)
+   2. Decode the signature into :math:`(\tilde{c}, z, h)` using ``decode_signature``
+   3. *Abort with "not valid"* if the Hamming weight of :math:`h > \omega`
```

Comment:
I don't see this condition in the standard?

Reply:
This is part of [the Dilithium round 3 spec](https://pq-crystals.org/dilithium/data/dilithium-specification-round3.pdf) (Section 3, Figure 4, line 31):

![image](https://github.com/user-attachments/assets/6fff9259-e9e8-4bbe-a452-f711d37d519c)

I'm guessing that it was removed during standardization. My guess is that it would reject signatures with hints revealing more than they should (?). But that's not the immediate problem of the verifier...

Oh, and I just saw that there's a remnant of this check in FIPS-204 Section 6.3 "ML-DSA Verifying (Internal)", where it reads in the last paragraph of the textual description:

> [...], the verifier checks that all of the coefficients of z are sufficiently small, **h contains no more than 𝜔 nonzero coefficients**, and c matches the hash c' [...]

In Botan, we have [a TODO that this check is not part of the standard anymore](https://github.com/randombit/botan/blob/dc326edf1204ec0875f807ffc475717d4c96410f/src/lib/pubkey/dilithium/dilithium_common/dilithium.cpp#L286-L290). But then, it actually kind of is still in the standard.

I'll add a Note describing the situation.

Reply:
Side note: the hint packing algorithms (FIPS-204 A. 20 + 21) are designed in a way that they cannot store a hint that has a hamming weight greater than 𝜔. Thus, using the standardized signature format, it isn't possible to create a signature that would ever fail this check. We should remove it from the implementation.